modules/pam_faillock/tst-pam_faillock-retval.c

```diff
@@ -0,0 +1,115 @@
+	ASSERT_EQ(0, fclose(fp));
+
+	/* root has access */
+	ASSERT_NE(NULL, fp = fopen(service_file, "w"));
+	ASSERT_LT(0, fprintf(fp, "#%%PAM-1.0\n"
+		  "auth [success=1 default=bad] %s/../pam_debug/.libs/pam_debug.so auth=success cred=success\n"
+		  "auth [default=die] %s/.libs/%s.so dir=%s authfail even_deny_root conf=%s\n"
+	          "auth sufficient %s/.libs/%s.so dir=%s authsucc even_deny_root conf=%s\n",
+	          cwd,
+		  cwd, MODULE_NAME, cwd, config_filename,
+	          cwd, MODULE_NAME, cwd, config_filename));
+
+	ASSERT_EQ(0, fclose(fp));
+
+	ASSERT_EQ(PAM_SUCCESS,
+		  pam_start_confdir(service_file, user_name, &conv, ".", &pamh));
+	ASSERT_NE(NULL, pamh);
+	ASSERT_EQ(PAM_SUCCESS, pam_authenticate(pamh, 0));
+	ASSERT_EQ(PAM_SUCCESS, pam_end(pamh, 0));
+	ASSERT_EQ(0, unlink(service_file));
+	pamh = NULL;
+
+	/* root tries to login 2 times without success*/
+	ASSERT_NE(NULL, fp = fopen(service_file, "w"));
+	ASSERT_LT(0, fprintf(fp, "#%%PAM-1.0\n"
+                  "auth requisite %s/.libs/%s.so dir=%s preauth even_deny_root conf=%s\n"
+		  "auth [success=1 default=bad] %s/../pam_debug/.libs/pam_debug.so auth=perm_denied cred=success\n"
+		  "auth [default=die] %s/.libs/%s.so dir=%s authfail even_deny_root conf=%s\n"
+	          "auth sufficient %s/.libs/%s.so dir=%s authsucc even_deny_root conf=%s\n",
+		  cwd, MODULE_NAME, cwd, config_filename,
+		  cwd,
+		  cwd, MODULE_NAME, cwd, config_filename,
+	          cwd, MODULE_NAME, cwd, config_filename));
+
+	ASSERT_EQ(0, fclose(fp));
+
+	ASSERT_EQ(PAM_SUCCESS,
+		  pam_start_confdir(service_file, user_name, &conv, ".", &pamh));
+	ASSERT_NE(NULL, pamh);
+	ASSERT_EQ(PAM_PERM_DENIED, pam_authenticate(pamh, 0));
+	ASSERT_EQ(PAM_PERM_DENIED, pam_authenticate(pamh, 0));
+	pamh = NULL;
+	ASSERT_EQ(0, unlink(service_file));
+
+	/* root is locked for 5 sec*/
+	ASSERT_NE(NULL, fp = fopen(service_file, "w"));
+	ASSERT_LT(0, fprintf(fp, "#%%PAM-1.0\n"
+                  "auth requisite %s/.libs/%s.so dir=%s preauth even_deny_root conf=%s\n"
+		  "auth [success=1 default=bad] %s/../pam_debug/.libs/pam_debug.so auth=success cred=success\n"
+		  "auth [default=die] %s/.libs/%s.so dir=%s authfail even_deny_root conf=%s\n"
+	          "auth sufficient %s/.libs/%s.so dir=%s authsucc even_deny_root conf=%s\n",
+		  cwd, MODULE_NAME, cwd, config_filename,
+		  cwd,
+	          cwd, MODULE_NAME, cwd, config_filename,
+	          cwd, MODULE_NAME, cwd, config_filename));
+
+	ASSERT_EQ(0, fclose(fp));
+
+	ASSERT_EQ(PAM_SUCCESS,
+		  pam_start_confdir(service_file, user_name, &conv, ".", &pamh));
+	ASSERT_NE(NULL, pamh);
+	ASSERT_EQ(PAM_AUTH_ERR, pam_authenticate(pamh, 0));
+
+	/* waiting at least 5 sec --> login is working again*/
+	sleep(6);
+	ASSERT_EQ(PAM_SUCCESS, pam_authenticate(pamh, 0));
+
+	ASSERT_EQ(PAM_SUCCESS, pam_end(pamh, 0));
+	ASSERT_EQ(0, unlink(service_file));
+	pamh = NULL;
+
+	ASSERT_LT(0,asprintf(&filename, "%s/%s", cwd, user_name));
+	ASSERT_EQ(0,unlink(filename));
+	free(filename);
```

Comment:
Yet again. filename doesn't have to be an absolute path, I suppose you can just do
```
ASSERT_EQ(0, unlink(user_name));
```
instead.